### ISSUE TITLE
Migrate Giraffa to Maven build infrastructure

### ISSUE THREAD

**Issue by shvachko** [GiraffaFS/giraffa#59]
Original [issue 59](https://code.google.com/p/giraffa/issues/detail?id=59) created by shvachko on 2013-01-10T09:59:09.000Z:

Currently Giraffa is using Ant as a build and dependency management tool. Ant is a great tool for some projects, however with Hadoop community migrating to Maven as a build build system. e.g. https://issues.apache.org/jira/browse/HADOOP-6671 Giraffa also needs some attention. 

Giraffa code base is not that big yet and it should be pretty easy to convert build.xml into pom.xml file and mavenize the project. Mavenization should facilitate latter integration of Giraffa into hadoop commons and will help Giraffa follow industry's best build practises.

**Comment by shvachko** [comment]
Comment [#1](https://code.google.com/p/giraffa/issues/detail?id=59#c1) originally posted by shvachko on 2013-01-10T18:28:00.000Z:

Yes please! Looking forward to this. Konstantin and I discussed ant vs maven a while back, no preference, we only went with ant because I knew it better than maven.

I would love to keep it up to date / closer to the Hadoop project and mavenization is definitely welcome.

**Comment by shvachko** [comment]
Comment [#2](https://code.google.com/p/giraffa/issues/detail?id=59#c2) originally posted by shvachko on 2013-01-10T19:17:13.000Z:

Sounds good! I'll work on this issue and it will give me a good chance to play around with the project's code base while I work on conversion.

**Comment by shvachko** [comment]
Comment [#3](https://code.google.com/p/giraffa/issues/detail?id=59#c3) originally posted by shvachko on 2013-01-10T19:41:15.000Z:

I usually come from a perspective of what is the immediate practical advantage of introducing a new technology. If e.g. it solves issue#&nbsp;4, even as a side effect, it is a justification for me. So what advantages do you see maven brings to Giraffa?

**Comment by shvachko** [comment]
Comment [#4](https://code.google.com/p/giraffa/issues/detail?id=59#c4) originally posted by shvachko on 2013-01-10T20:09:13.000Z:

Yes, in maven it's possible to generate different distributions based on specific profile so issue#&nbsp;4 will be a side effect of issue#&nbsp;59. When I design maven project structure I always try to keep clear separation between project configuration such as distro or packaging and code itself. Usually it makes project structure and future modification and support much easier.

**Comment by shvachko** [comment]
Comment [#5](https://code.google.com/p/giraffa/issues/detail?id=59#c5) originally posted by shvachko on 2013-01-10T22:05:31.000Z:

Yes good point. We haven't discussed issue #&nbsp;4 in a long time. I can certainly agree.

My own advantages that I see are that we become closer to the Hadoop community; making it easier (hopefully) for current Hadoop developers / committers to take a look at Giraffa and begin diving in.
 I have seen many project battle over ant vs. maven and eventually go maven; I think it would behoove us to skip any drama and just let maven take the course.
Maven also appears to provide a lot of abstraction and simplicity to the compilation process, which can be daunting at times.

Personally, I would also like to LEARN maven so I can understand what is going on under that 'mvn install' command.

**Comment by shvachko** [comment]
Comment [#6](https://code.google.com/p/giraffa/issues/detail?id=59#c6) originally posted by shvachko on 2013-01-10T23:15:44.000Z:

Let's do maven then.
Konstantin I added you to contributors. You should be able to assign the issue to yourself. Let me know if there are any problems.

**Comment by shvachko** [comment]
Comment [#7](https://code.google.com/p/giraffa/issues/detail?id=59#c7) originally posted by shvachko on 2013-01-10T23:26:45.000Z:

Konstantin, seems that I can't modify Owner field

**Comment by shvachko** [comment]
Comment [#8](https://code.google.com/p/giraffa/issues/detail?id=59#c8) originally posted by shvachko on 2013-01-10T23:48:18.000Z:

Just saving our email conversation here:

PJ> Honestly right now most of the work is just patching; if you think
PJ> mavenization will require a separate branch than by all means do as
PJ> you wish.

PJ> Most likely we would have to do some sort of code review or something
PJ> at that point -- otherwise though I am assuming most of the work would
PJ> be in generating a single pom.xml file, correct?
PJ> If we only need that file, then perhaps it is best to just submit it
PJ> as a patch and we can +1 it and remove the ant code ourselves.

KP> Mavenization will need a bit more than just one file. Maven follows some 
KP> specific file structure convention, so a some files will need to get 
KP> moved around. Tests will need to go to its own location.

KP> On this stage of the project it's pretty simple to convert to Maven, 
KP> but once projects grows beyond some point, it becomes pain, so it's 
better to do earlier than later.

KP> I'll also write detailed documentation on how to setup maven environment 
KP> so appropriate sections of the website can be updated

**Comment by shvachko** [comment]
Comment [#9](https://code.google.com/p/giraffa/issues/detail?id=59#c9) originally posted by shvachko on 2013-01-10T23:50:51.000Z:

I will update the field for him. Konst, can we change contributor permissions to set / update owner field? I know I cannot.

Also yes, continuing from email my vote is on a separate branch then. We can merge it in once it is ready / proven. What do you think Konst? Will we do a code review once it is ready?

**Comment by shvachko** [comment]
Comment [#10](https://code.google.com/p/giraffa/issues/detail?id=59#c10) originally posted by shvachko on 2013-01-10T23:54:04.000Z:

&lt;empty&gt;

**Comment by shvachko** [comment]
Comment [#11](https://code.google.com/p/giraffa/issues/detail?id=59#c11) originally posted by shvachko on 2013-01-11T05:09:15.000Z:

I think I updated permissions. Kostya could you please try to update Type-Defect to Type-Enhancement or something.
Another thing that I noticed that we almost exhausted out attachments limit. Should we clean up some big ones in closed issues?

**Comment by shvachko** [comment]
Comment [#12](https://code.google.com/p/giraffa/issues/detail?id=59#c12) originally posted by shvachko on 2013-01-11T05:12:24.000Z:

On the branches. You will probably work on a local branch for mavenization. If you find changes too large for a single patch we can create a new dev branch, no problem.

**Comment by shvachko** [comment]
Comment [#13](https://code.google.com/p/giraffa/issues/detail?id=59#c13) originally posted by shvachko on 2013-01-11T05:57:33.000Z:

&lt;empty&gt;

**Comment by shvachko** [comment]
Comment [#14](https://code.google.com/p/giraffa/issues/detail?id=59#c14) originally posted by shvachko on 2013-01-11T06:00:04.000Z:

Konstantin, looks like permission are OK now. I'm able to change fields and assign owner. I will work on mavenization in local branch for now and will evaluate amount of changes latter when its ready for push.

**Comment by shvachko** [comment]
Comment [#15](https://code.google.com/p/giraffa/issues/detail?id=59#c15) originally posted by shvachko on 2013-01-11T16:49:14.000Z:

Cool!

**Comment by shvachko** [comment]
Comment [#16](https://code.google.com/p/giraffa/issues/detail?id=59#c16) originally posted by shvachko on 2013-01-11T20:07:02.000Z:

&lt;empty&gt;

**Comment by shvachko** [comment]
Comment [#17](https://code.google.com/p/giraffa/issues/detail?id=59#c17) originally posted by shvachko on 2013-01-11T20:08:57.000Z:

Konst, there should be free space in attachments. I deleted some of the massive attachments I had back during the TestGiraffaCLI issue.

Kosta, please ignore the "blocking" issues, they should really be "related to" but there is no such field available.

**Comment by shvachko** [comment]
Comment [#18](https://code.google.com/p/giraffa/issues/detail?id=59#c18) originally posted by shvachko on 2013-01-21T08:18:10.000Z:

Just wanted to update everybody that I went though every line of code and ran all the tests. Everything seems to work without a problem. I'll be starting on maven conversion tomorrow.

**Comment by shvachko** [comment]
Comment [#19](https://code.google.com/p/giraffa/issues/detail?id=59#c19) originally posted by shvachko on 2013-01-24T08:08:44.000Z:

I'm almost done with conversion and I did a minor re-factoring of test files. After refactoring I couldn't get TestGiraffaFileStatus to pass in testCompletedLocatedFileStatus() method because I replaced the following line 

assert blocks.getLocatedBlocks().size() == (int) Math.ceil(9876/512) :
       "Incorrect number of blocks returned.";

with

assertEquals("Incorrect number of blocks returned.", (int)Math.ceil(9876/512d), blocks.getLocatedBlocks().size());

and it was failing with 

java.lang.AssertionError: Incorrect number of blocks returned. expected:<19> but was:<20> at org.junit.Assert.fail(Assert.java:91)

so looks like original assert statement had a typo and didn't use standard "assertEquals" method and skipped actual comparison of the values, therefore it always passed. Also division 9876/512 is rounded to 19 before it gets converted to double, so Math.ceil always returns 19 instead of 20, which would be revealed if assert form was correct.

**Comment by shvachko** [comment]
Comment [#20](https://code.google.com/p/giraffa/issues/detail?id=59#c20) originally posted by shvachko on 2013-01-24T19:41:06.000Z:

We need to get a license for clover plugin. It's free for open source project. Konstantin, can you please fill out this form and forward me the license when you get it, so that I could integrate it into the build. 

http://www.atlassian.com/software/views/open-source-license-request

Thanks!

**Comment by shvachko** [comment]
Comment [#21](https://code.google.com/p/giraffa/issues/detail?id=59#c21) originally posted by shvachko on 2013-01-24T20:16:49.000Z:

I believe Konstantin already has a Clover license as part of his Apache committer status. It may be a good idea to get one for ourselves though. Konst, if you want I can do the form as well, up to you.

**Comment by shvachko** [comment]
Comment [#22](https://code.google.com/p/giraffa/issues/detail?id=59#c22) originally posted by shvachko on 2013-01-25T02:21:57.000Z:

So I'm done with 59. How do you want to proceed? Since it includes quite a lot of changes I propose to create a new branch "maven" and push all changes to "maven" branch, then you will review it and merge into master if everything is ok. Afterwards "maven" branch can be deleted.

**Comment by shvachko** [comment]
Comment [#23](https://code.google.com/p/giraffa/issues/detail?id=59#c23) originally posted by shvachko on 2013-01-25T02:54:40.000Z:

By the way, if you use hudson or jenkins we can set up maven build to automatically deploy Report Site to github. For example: http://kpelykh.github.com/giraffa

**Comment by shvachko** [comment]
Comment [#25](https://code.google.com/p/giraffa/issues/detail?id=59#c25) originally posted by shvachko on 2013-01-30T00:50:50.000Z:

Sorry for a dup message - used wrong account first time.

Attached is a diff file that fixes issue59 and converts environment to Maven. 
It can also build website that includes multiple reports such as clover, checkstyle, firebugs.. 

README.txt file contains instructions on how to build the project. 

To apply the patch:

"git diff --ignore-space-change --ignore-whitespace issue59.diff" 

After applying the patch you should be able run maven build including tests:

"mvn clean install"

If build was successful,  you will see similar output:

[INFO] Installing /home/konstantin/Development/GiraffaProject/giraffa/target/hadoop-hdfs-giraffa-0.0.1-SNAPSHOT.jar to /home/konstantin/.m2/repository/org/apache/hadoop/hadoop-hdfs-giraffa/0.0.1-SNAPSHOT/hadoop-hdfs-giraffa-0.0.1-SNAPSHOT.jar
[INFO] Installing /home/konstantin/Development/GiraffaProject/giraffa/pom.xml to /home/konstantin/.m2/repository/org/apache/hadoop/hadoop-hdfs-giraffa/0.0.1-SNAPSHOT/hadoop-hdfs-giraffa-0.0.1-SNAPSHOT.pom
[INFO] Installing /home/konstantin/Development/GiraffaProject/giraffa/target/hadoop-hdfs-giraffa-0.0.1-SNAPSHOT-sources.jar to /home/konstantin/.m2/repository/org/apache/hadoop/hadoop-hdfs-giraffa/0.0.1-SNAPSHOT/hadoop-hdfs-giraffa-0.0.1-SNAPSHOT-sources.jar
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 6:28.285s
[INFO] Finished at: Tue Jan 29 15:22:29 PST 2013
[INFO] Final Memory: 33M/288M
[INFO] ------------------------------------------------------------------------

**Comment by shvachko** [comment]
Comment [#26](https://code.google.com/p/giraffa/issues/detail?id=59#c26) originally posted by shvachko on 2013-01-30T02:54:27.000Z:

This looks really good. I will need to set aside some time this week to play around with your work. :) 
Thank you for our first major contribution Kosta!

**Comment by shvachko** [comment]
Comment [#27](https://code.google.com/p/giraffa/issues/detail?id=59#c27) originally posted by shvachko on 2013-01-31T02:05:24.000Z:

Looking good! I ran the build and tests and looking at the clover report right now. This will look lovely on our Jenkins build as well; which is hosted by CloudBees @ here: https://giraffa.ci.cloudbees.com/job/Giraffa-TRUNK/

Notes:

1) There are some folders that do not get deleted from the patch. I take it this is because you cannot specify directories to be deleted in a patch file. So it is fine.

2) We are still suffering from some OoM exception somewhere; it is causing the tests to fail in their current state. Do you see the tests failing as well, Kosta? What is your work environment?

3) I really like generating the site and clover reports. +1 there.

4) Where are the dependency variables stored in Maven? Like ${hadoop.common.version} and the others. I want to experiment with upgrading dependencies.

**Comment by shvachko** [comment]
Comment [#28](https://code.google.com/p/giraffa/issues/detail?id=59#c28) originally posted by shvachko on 2013-01-31T02:45:55.000Z:

Also a tiny nit -- we can use this patch to remove Ivy as well. Clearly Maven will handle our dependencies now.

**Comment by shvachko** [comment]
Comment [#29](https://code.google.com/p/giraffa/issues/detail?id=59#c29) originally posted by shvachko on 2013-01-31T03:46:45.000Z:

Here's my settings: 

export MAVEN_OPTS="-Xmx2g -XX:MaxPermSize=512m"

With this settings I never got OOM error during maven builds. 

All the dependency versions stored in property section of pom.xml file. Usually if you have multi module project then all the properties are moved out to so called parent pom file, but since in our case we have just one pom.xml I decided to keep everything locally in the same pom. 

Here's folder structure for maven project, all other folders can be removed:

.
 |-src
 |---main
 |-----java
 |-------org
 |---------apache
 |-----------giraffa
 |-------------hbase
 |-----------hadoop
 |-------------hdfs
 |-----resources
 |-------scripts
 |---site
 |---test
 |-----clover
 |-----java
 |-------org
 |---------apache
 |-----------giraffa
 |-----resources

**Comment by shvachko** [comment]
Comment [#30](https://code.google.com/p/giraffa/issues/detail?id=59#c30) originally posted by shvachko on 2013-01-31T04:02:33.000Z:

The only dependency versions which is specified right now are 

 <hadoop.version>0.22.0</hadoop.version>
 <hbase.version>0.94.1</hbase.version>

to get full tree of dependencies run this command from project folder:

$ org.apache.maven.plugins:maven-dependency-plugin:2.6:tree

you will get the following output:

[INFO] org.apache.hadoop:hadoop-hdfs-giraffa:jar:0.0.1-SNAPSHOT
[INFO] +- org.apache.hadoop:hadoop-hdfs:jar:0.22.0:compile
[INFO] +- org.apache.hadoop:hadoop-common:jar:0.22.0:compile
[INFO] |  +- commons-cli:commons-cli:jar:1.2:compile
[INFO] |  +- xmlenc:xmlenc:jar:0.52:compile
[INFO] |  +- commons-codec:commons-codec:jar:1.4:compile
[INFO] |  +- commons-logging:commons-logging:jar:1.1.1:compile
[INFO] |  +- org.slf4j:slf4j-api:jar:1.6.1:compile
....

all those versions come from 

hadoop-common pom file, you can view this file here 

http://repo1.maven.org/maven2/org/apache/hadoop/hadoop-common/0.22.0/hadoop-common-0.22.0.pom

also I you can use this website to search for all available maven artifacts

http://mvnrepository.com

e.g. http://mvnrepository.com/artifact/org.apache.hadoop/hadoop-common/0.22.0

if you want to override any of dependencies or exclude some you need to declare it in <dependencyManagement> section. for example to if I want to replace org.slf4j:slf4j-api:jar:1.6.1 with 1.7.2 then I would add this section into pom.xml:

<dependencyManagement>
   <dependencies>
      <dependency>
         <groupId>org.slf4j</groupId>
         <artifactId>slf4j-api</artifactId>
         <version>1.7.2</version>
      </dependency>
   </dependencies>
</dependencyManagement>

Then if you run 

$ mvn org.apache.maven.plugins:maven-dependency-plugin:2.6:tree

you would see 

[INFO] |  +- org.slf4j:slf4j-api:jar:1.7.2:compile (version managed from 1.6.1)

**Comment by shvachko** [comment]
Comment [#31](https://code.google.com/p/giraffa/issues/detail?id=59#c31) originally posted by shvachko on 2013-02-07T03:36:00.000Z:

I've been waiting on Konst before committing. Konst, can you please check out the patch and confirm it working for you?

Sorry its taken so long. We both got swamped with work and other stuff.
Thank you for this contribution. Really. Let's push it through!

**Comment by shvachko** [comment]
Comment [#32](https://code.google.com/p/giraffa/issues/detail?id=59#c32) originally posted by shvachko on 2013-02-07T06:45:34.000Z:

Plamen, no problem, I totally understand that it takes a bit longer sometimes, considering that we all have day jobs. Once Maven patch is in place I will pick another issue to work on..

**Comment by shvachko** [comment]
Comment [#33](https://code.google.com/p/giraffa/issues/detail?id=59#c33) originally posted by shvachko on 2013-02-08T19:48:42.000Z:

Starting to catch up on the giraffa issues. Sorry for long delay.

**Comment by shvachko** [comment]
Comment [#34](https://code.google.com/p/giraffa/issues/detail?id=59#c34) originally posted by shvachko on 2013-02-09T09:51:58.000Z:

No problem. Please remember that this issue takes precedence over committing your new issues however because of the reliance of the patch on revision 63dd1dc7567a.

**Comment by shvachko** [comment]
Comment [#35](https://code.google.com/p/giraffa/issues/detail?id=59#c35) originally posted by shvachko on 2013-02-09T11:19:15.000Z:

&lt;empty&gt;

**Comment by shvachko** [comment]
Comment [#36](https://code.google.com/p/giraffa/issues/detail?id=59#c36) originally posted by shvachko on 2013-02-12T06:50:11.000Z:

Kostya, the patch looks great. Giraffa compiles and runs tests. Two things: 
1. Don't know if it was intended, but the tests logging goes directly to std output, which is not very convenient. Can we redirect it to some log file under target/
2. We cannot apply the patch as a direct diff. Because we will loose the history of the individual files. We need to use renames under git and them commit new maven files. What do you think? 
Also that way the changes to the source tree will not effect the mavenization procedure.

**Comment by shvachko** [comment]
Comment [#37](https://code.google.com/p/giraffa/issues/detail?id=59#c37) originally posted by shvachko on 2013-02-12T07:31:10.000Z:
1. It will be trivial to redirect test logging to the file: 
   http://maven.apache.org/surefire/maven-surefire-plugin/test-mojo.html#redirectTestOutputToFile

I can do this either a part of issue59 or as a separate issue when mavenization if done. 

I'm not positive that Jenkins can tail file to show it in web ui, as opposed to std stream. I know that by default std stream is used. For jenkins we can use a separate profile which will instruct test pluging to stream output to std, if needed. 

2) That's correct -- applying patch directly will wipe out files' history. One way would be to create a separate branch (mvn) in git and to perform renames, then merge mvn branch into main branch. This way we will be able to keep history of every single file and information about file transformations.

**Comment by shvachko** [comment]
Comment [#38](https://code.google.com/p/giraffa/issues/detail?id=59#c38) originally posted by shvachko on 2013-02-12T08:39:01.000Z:

For (1), we can do it after mavenization.
For (2), can you just write down here the renaming steps and make a patch for new files only. I'll do a dry run on local branch. Then will just commit the changes.
Do you mean creating a remote branch "mvn"? Not sure how it helps.

**Comment by shvachko** [comment]
Comment [#39](https://code.google.com/p/giraffa/issues/detail?id=59#c39) originally posted by shvachko on 2013-02-12T19:26:38.000Z:

Regarding 2: Will do renaming today and will submit a new patch.
Yes, I was referring to creating a new remote branch "mvn" and then merging it with trunk. Please ignore it, since patch with renames is much easier and better way.

**Comment by shvachko** [comment]
Comment [#40](https://code.google.com/p/giraffa/issues/detail?id=59#c40) originally posted by shvachko on 2013-02-13T09:50:20.000Z:

Konstantin, please look at the attached patch. It was created with git format-patch command and contains meta information about all file manipulations (rename, delete, create). 

to apply the patch use:

git am --ignore-space-change --ignore-whitespace 0001-Issue-59-Migrate-Giraffa-to-Maven-build-infrastructu.patch

This will create a new commit, which will keep history of all files.

After applying patch, you should be able to run "mvn clean install"

**Comment by shvachko** [comment]
Comment [#42](https://code.google.com/p/giraffa/issues/detail?id=59#c42) originally posted by shvachko on 2013-02-13T09:53:10.000Z:

&lt;empty&gt;

**Comment by shvachko** [comment]
Comment [#43](https://code.google.com/p/giraffa/issues/detail?id=59#c43) originally posted by shvachko on 2013-02-16T23:12:22.000Z:

Trying to apply the patch using your command and getting "does not match index" error. See the message below.
My tree is clean. Did you format your patch with a previous version or in a local branch?
How can I avoid the error?

Applying: Issue 59: Migrate Giraffa to Maven build infrastructure (Konstantin P)
error: build.xml: does not match index
error: conf/giraffa-site.xml: does not match index
error: src/java/org/apache/giraffa/DirectoryTable.java: does not match index
error: src/java/org/apache/giraffa/FileField.java: does not match index
error: src/java/org/apache/giraffa/FullPathRowKey.java: does not match index
error: src/java/org/apache/giraffa/GiraffaConfiguration.java: does not match index
error: src/java/org/apache/giraffa/GiraffaConstants.java: does not match index
error: src/java/org/apache/giraffa/GiraffaFileSystem.java: does not match index
error: src/java/org/apache/giraffa/INode.java: does not match index
error: src/java/org/apache/giraffa/RowKey.java: does not match index
error: src/java/org/apache/giraffa/hbase/BlockManagementAgent.java: does not match index
error: src/java/org/apache/giraffa/hbase/NamespaceAgent.java: does not match index
error: src/java/org/apache/hadoop/hdfs/GiraffaClient.java: does not match index
error: src/java/org/apache/giraffa/giraffa-default.xml: does not match index
error: bin/giraffa: does not match index
error: bin/giraffa-config.sh: does not match index
error: bin/start-giraffa.sh: does not match index
error: bin/stop-giraffa.sh: does not match index
error: src/test/org/apache/giraffa/GiraffaTestUtils.java: does not match index
error: src/test/org/apache/giraffa/TestBlockManagement.java: does not match index
error: src/test/org/apache/giraffa/TestDirectoryTable.java: does not match index
error: src/test/org/apache/giraffa/TestGiraffaFS.java: does not match index
error: src/test/org/apache/giraffa/TestGiraffaFSNegative.java: does not match index
error: src/test/org/apache/giraffa/TestRestartGiraffa.java: does not match index
Patch failed at 0001 Issue 59: Migrate Giraffa to Maven build infrastructure (Konstantin P)
When you have resolved this problem run "git am --resolved".
If you would prefer to skip this patch, instead run "git am --skip".
To restore the original branch and stop patching run "git am --abort".

**Comment by shvachko** [comment]
Comment [#44](https://code.google.com/p/giraffa/issues/detail?id=59#c44) originally posted by shvachko on 2013-02-17T01:37:09.000Z:

OK nm. I did git reset --hard and it applied.

**Comment by shvachko** [comment]
Comment [#45](https://code.google.com/p/giraffa/issues/detail?id=59#c45) originally posted by shvachko on 2013-02-17T03:10:22.000Z:

I was trying to test the history after applying with
git log src/main/java/org/apache/giraffa/RowKey.java
And was getting only the last commit. Found out now that in order to get full history use --follow option:

git log --follow src/main/java/org/apache/giraffa/RowKey.java

All looks good to me, modular logging to stdout.
+1 will commit now.

**Comment by shvachko** [comment]
Comment [#46](https://code.google.com/p/giraffa/issues/detail?id=59#c46) originally posted by shvachko on 2013-02-17T03:35:10.000Z:

I just committed this. Thank you Kostya.
Congratulations to your first Giraffa issue!

**Comment by shvachko** [comment]
Comment [#47](https://code.google.com/p/giraffa/issues/detail?id=59#c47) originally posted by shvachko on 2013-02-19T01:40:36.000Z:

Issue 23 has been merged into this issue.

**Comment by shvachko** [comment]
Comment [#48](https://code.google.com/p/giraffa/issues/detail?id=59#c48) originally posted by shvachko on 2013-02-19T07:45:10.000Z:

Issue 26 has been merged into this issue.
